specs/APIReview_CustomDownload.md

```diff
@@ -0,0 +1,346 @@
+    [uuid(34acb11c-fc37-4418-9132-f9c21d1eafb9), object, pointer_default(unique)]
+    interface ICoreWebView2NewWindowRequestedEventArgs : IUnknown
+    {
+        // ...
+
+        /// Window features specified by the window.open call.
+        /// These features can be considered for positioning and sizing of
+        /// new webview windows.
+        [propget] HRESULT WindowFeatures([out, retval] ICoreWebView2WindowFeatures** windowFeatures);
+    }
+    ```
+
+    ```c# (but really MIDL3)
+    public class CoreWebView2NewWindowRequestedEventArgs
+    {
+        // ...
+
+	       public CoreWebView2WindowFeatures WindowFeatures { get; }
+    }
+    ```
+-->
+## Win32 C++
+```c#
+[uuid(9EAFB7D0-88C3-4450-BBFB-C05A46C40C72), object, pointer_default(unique)]
+interface ICoreWebView2Staging2 : IUnknown {
+
+  /// Add an event handler for the `DownloadStarting` event. This event fires
+  /// when a download has begun, blocking the default download UI.
+  ///
+  /// The host can choose to cancel a download, change the save path, and hide
+  /// the default download UI. If the host chooses to cancel the download, the
+  /// download is not saved and no UI is shown. Otherwise, the download is saved
+  /// once the event completes, and default download UI is shown if the host
+  /// did not choose to hide it.
+  ///
+  /// If the event is not handled, downloads will complete normally with the
+  /// default UI shown.
+  ///
+  HRESULT add_DownloadStarting(
+    [in] ICoreWebView2StagingDownloadStartingEventHandler* eventHandler,
+    [out] EventRegistrationToken* token);
+
+  /// Remove an event handler previously added with `add_DownloadStarting`.
+  HRESULT remove_DownloadStarting(
+      [in] EventRegistrationToken token);
+}
+
+[uuid(53b676ec-c3b1-4804-996a-c72c16b09efb), object, pointer_default(unique)]
+interface ICoreWebView2StagingDownloadStartingEventHandler : IUnknown
+{
+  /// Provides the event args for the corresponding event.
+  HRESULT Invoke(
+      [in] ICoreWebView2Staging2* sender,
+      [in] ICoreWebView2StagingDownloadStartingEventArgs* args);
+}
+
+/// Event args for the `DownloadStarting` event.
+[uuid(c7e95e2f-f789-44e6-b372-3141469240e4), object, pointer_default(unique)]
+interface ICoreWebView2StagingDownloadStartingEventArgs : IUnknown
+{
+  /// The uri of the download.
+  [propget] HRESULT Uri([out, retval] LPWSTR* uri);
+
+  /// The Content-Disposition header value from HTTP response.
+  [propget] HRESULT ContentDisposition([out, retval] LPWSTR* contentDisposition);
+
+  /// MIME type of the downloaded content.
+  [propget] HRESULT MimeType([out, retval] LPWSTR* mimeType);
+
+  /// The total number of expected bytes.
+  [propget] HRESULT TotalBytes([out, retval] UINT64* totalBytes);
+
+  /// The host may set this flag to cancel the download.
+  [propget] HRESULT Cancel([out, retval] BOOL* cancel);
+
+  /// Sets the `Cancel` property.
+  [propput] HRESULT Cancel([in] BOOL cancel);
+
+  /// The save path of the download.
```

Comment:
Is this supposed to be the path to the folder into which to put the file, or the path to the file? If its one of those, what happens if you provide the other?

Reply:
The path to the file. We can choose to implement it in such a way that we use the default filename if only the folder path is given? Related to this, the chrome.downloads extension API only allows paths relative to the Downloads folder. Not sure if that's something we want to do https://developer.chrome.com/docs/extensions/reference/downloads/#type-FilenameSuggestion

Reply:
Sorry not trying to suggest a change in behavior necessarily. Just that we need to document what the property is exactly (the path to the file sounds fine) and what happens if the thing provided is not a valid file path, or the containing folder doesn't exist yet or those kinds of things.